### PR TITLE
C# Message Routing Sample

### PR DESCRIPTION
**Pull request by carlosscastro**

### CODE REVIEW

File: csharp_dotnetcore/9.Message-Routing/BotServices.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using Microsoft.Bot.Builder.AI.Luis;
+using Microsoft.Bot.Builder.AI.QnA;
+
+namespace MessageRoutingBot
+{
+    /// <summary>
+    /// Class represents bot configuration.
+    /// </summary>
+    [Serializable]
+    public class BotServices
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/BotServices.cs]: Remove this attribute.

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    /// </summary>
+    public class MessageRoutingBot : IBot
+    {
+        private readonly BotServices _services;
+        private readonly SemaphoreSlim _semaphore;
+        private readonly MessageRoutingBotAccessors _accessors;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnterpriseBot"/> class.
+        /// </summary>
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: EnterpriseBot -> MessageRoutingBot

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        /// <param name="botServices">Bot services.</param>
+        /// <param name="accessors">Bot State Accessors.</param>
+        public MessageRoutingBot(BotServices botServices, MessageRoutingBotAccessors accessors)
+        {
+            _accessors = accessors;
+            _services = botServices;
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: null checks. Both these parms are required.

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+            // a semaphore to serialize access to the bot state
+            _semaphore = accessors.SemaphoreSlim;
+
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: eek!

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        /// </summary>
+        /// <param name="turnContext">Bot Turn Context.</param>
+        /// <param name="cancellationToken">Task CancellationToken.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await _semaphore.WaitAsync();
+
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: It's far from obvious what this is doing. I'm also not convinced it's right. 

We want to (if I understand correctly) serialize on a per-conversation basis. Got globally serialized.

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.csproj [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    <PackageReference Include="Microsoft.Bot.Builder.Azure" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Dialogs" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.Integration.AspNet.Core" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Builder.TemplateManager" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Configuration" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Connector" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Bot.Schema" Version="4.0.0.38762" />
+    <PackageReference Include="Microsoft.Graph" Version="1.10.0" />
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.0-beta008">
+      <PrivateAssets>all</PrivateAssets>
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.csproj]: Missing the Async Analyzers here as well.

File: csharp_dotnetcore/9.Message-Routing/Program.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.AspNetCore;
+using Microsoft.AspNetCore.Hosting;
+
+namespace MessageRoutingBot
+{
+    /// <summary>
+    /// Entry point for Bot application.
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/Program.cs]: Remove comments from this class, in alignment with what the ASP.Net Core team does.

File: csharp_dotnetcore/9.Message-Routing/Startup.cs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                // based storage providers, add the Microsoft.Bot.Builder.Azure
+                // Nuget package to your solution. That package is found at:
+                // https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/
+                // Uncomment this line to use Azure Blob Storage
+                // IStorage dataStore = new Microsoft.Bot.Builder.Azure.AzureBlobStorage("AzureBlobConnectionString", "containerName");
+                // Create and add conversation state.
+                var convoState = new ConversationState(dataStore);
+                options.State.Add(convoState);
+                options.Middleware.Add(new BotStateSet(options.State.ToArray()));
+            });
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/Startup.cs]: Remove BotStateSet in alignment with our other samples. Just call "Save" in the main onTurn method. Makes things much simpler acorss the board, and developers can handle save errors if they happen.

File: csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.AI.Luis;
+using Newtonsoft.Json;
+using System.Collections.Generic;
+
+namespace Luis
+{
+    public class MessageRoutingBot_General : IRecognizerConvert
+    {
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs]: Comments. 
Also, weird namespace.

File: csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            ConfirmNo,
+            ConfirmMore,
+            Next,
+            Goodbye,
+            None
+        };
+        public Dictionary<Intent, IntentScore> Intents;
+
+        public class _Entities
+        {
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs]: Lots of interesting things here. No comments, 

Also, what's with the "_" naming convention?

File: csharp_dotnetcore/9.Message-Routing/Dialogs/Cancel/CancelResponses.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+namespace MessageRoutingBot
+{
+    /// <summary>
+    /// Responses for the <see cref="CancelDialog"/>.
+    /// </summary>
+    public class CancelResponses : TemplateManager
+    {
+        // Constants
+        public const string _confirmPrompt = "Cancel.ConfirmCancelPrompt";
+        public const string _cancelConfirmed = "Cancel.CancelConfirmed";
```
Comment (cleemullins) [inline csharp_dotnetcore/9.Message-Routing/Dialogs/Cancel/CancelResponses.cs]: public const shouldn't have "_" and should start with a Capital Letter

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs (hunk 4 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: I'll remove this ha.

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs (hunk 5 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.cs]: We will remove this. When this was started, this was one of the patterns we were considering for the enterprise bot but not anymore!

File: csharp_dotnetcore/9.Message-Routing/Program.cs (hunk 7 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/Program.cs]: Sounds good.

File: csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs (hunk 10 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs]: @cleemullins this is generated by luisgen, as stated in the upper part. Yes, it does not match our patterns :(

File: csharp_dotnetcore/9.Message-Routing/Dialogs/Cancel/CancelResponses.cs (hunk 11 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/Dialogs/Cancel/CancelResponses.cs]: will fix!

File: csharp_dotnetcore/9.Message-Routing/Startup.cs (hunk 8 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/Startup.cs]: @cleemullins do you have an example of this? Is this save the one you mean?
https://github.com/Microsoft/BotBuilder-Samples/blob/v4/csharp_dotnetcore/5.MultiTurn-Prompts-Bot/MultiTurnPromptsBot.cs#L96

File: csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs (hunk 9 above)
Comment (carlosscastro) [reply csharp_dotnetcore/9.Message-Routing/CognitiveModels/LUIS/General.cs]: This is auto-generated by luis gen

File: csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.csproj (hunk 6 above)
Comment (cleemullins) [reply csharp_dotnetcore/9.Message-Routing/MessageRoutingBot.csproj]: I still don't see the async usage analyzer in here.. 

---
In reply to: [216445553](https://github.com/Microsoft/BotBuilder-Samples/pull/349#discussion_r216445553) [](ancestors = 216445553)

File: csharp_dotnetcore/9.Message-Routing/Startup.cs (hunk 8 above)
Comment (cleemullins) [reply csharp_dotnetcore/9.Message-Routing/Startup.cs]: Yes, that one is fine. Should take about 2 mintues to make this change... 

---
In reply to: [216499197](https://github.com/Microsoft/BotBuilder-Samples/pull/349#discussion_r216499197) [](ancestors = 216499197)

**Comment by carlosscastro** [comment]
Thanks for the review @cleemullins I addressed all your comments 💯 :shipit:

**Pull request merged**